File: webui/frontend/src/lib/personaParse.ts
/**
 * Static parse of openai-agents persona definitions (REQ-81 / #433).
 *
 * Mirrors `swarm.core.persona_parse`. Never executes source. Variable names
 * are skipped — unknown / unparsable source is count 1 and no invented names.
 */

export interface BlueprintPersona {
  name: string
}

export interface PersonaParseResult {
  count: number
  personas: BlueprintPersona[]
  parsed: boolean
}

export const UNPARSED_PERSONAS: PersonaParseResult = {
  count: 1,
  personas: [],
  parsed: false,
}

const PERSONA_CTORS = new Set(['Agent', 'make_agent', '_make_agent'])

const CALL_RE =
  /\b(Agent|make_agent|_make_agent)\s*\(([\s\S]*?)\)/g
const NAME_KW_RE = /(?:^|[,(\s])name\s*=\s*(['"])((?:\\.|(?!\1).)*)\1/
const FIRST_STR_RE = /^\s*(['"])((?:\\.|(?!\1).)*)\1/

export function parseOpenaiAgentPersonas(source: string | null | undefined): PersonaParseResult {
  if (typeof source !== 'string' || !source.trim()) return { ...UNPARSED_PERSONAS }
  // Mirror Python ast.parse: junk around an Agent(...) call is unparsed, not a roster.
  if (!pythonSourceLooksParseable(source)) return { ...UNPARSED_PERSONAS }
  const names: string[] = []
  const seen = new Set<string>()
  CALL_RE.lastIndex = 0
  let match: RegExpExecArray | null
  while ((match = CALL_RE.exec(source)) !== null) {
    const ctor = match[1]
    if (!ctor || !PERSONA_CTORS.has(ctor)) continue
    const args = match[2] || ''
    const name = literalName(args)
    if (!name || seen.has(name)) continue
    seen.add(name)
    names.push(name)
  }
  if (!names.length) return { ...UNPARSED_PERSONAS }
  return {
    count: names.length,
    personas: names.map((name) => ({ name })),
    parsed: true,
  }
}

/** Conservative stand-in for `ast.parse`. Unbalanced delimiters → unparsed. */
function pythonSourceLooksParseable(source: string): boolean {
  let paren = 0
  let bracket = 0
  let brace = 0
  let quote: '"' | "'" | null = null
  let triple = false
  let i = 0
  while (i < source.length) {
    const ch = source[i]
    const next2 = source.slice(i, i + 3)
    if (quote) {
      if (ch === '\\') {
        i += 2
        continue
      }
      if (triple) {
        if (next2 === quote + quote + quote) {
          quote = null
          triple = false
          i += 3
          continue
        }
      } else if (ch === quote) {
        quote = null
      }
      i += 1
      continue
    }
    if (ch === '#') {
      while (i < source.length && source[i] !== '\n') i += 1
      continue
    }
    if ((ch === '"' || ch === "'") && next2 === ch + ch + ch) {
      quote = ch
      triple = true
      i += 3
      continue
    }
    if (ch === '"' || ch === "'") {
      quote = ch
      i += 1
      continue
    }
    if (ch === '(') paren += 1
    else if (ch === ')') paren -= 1
    else if (ch === '[') bracket += 1
    else if (ch === ']') bracket -= 1
    else if (ch === '{') brace += 1
    else if (ch === '}') brace -= 1
    if (paren < 0 || bracket < 0 || brace < 0) return false
    i += 1
  }
  return paren === 0 && bracket === 0 && brace === 0 && quote === null
}

function literalName(args: string): string | null {
  const kw = args.match(NAME_KW_RE)
  if (kw?.[2]?.trim()) return kw[2].trim()
  const first = args.match(FIRST_STR_RE)
  if (first?.[2]?.trim()) return first[2].trim()
  return null
}

export function normalizePersonaResult(raw: unknown): PersonaParseResult {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return { ...UNPARSED_PERSONAS }
  const rec = raw as Record<string, unknown>
  const personas: BlueprintPersona[] = []
  const list = Array.isArray(rec.personas) ? rec.personas : []
  for (const item of list) {
    if (!item || typeof item !== 'object') continue
    const name = typeof (item as { name?: unknown }).name === 'string'
      ? (item as { name: string }).name.trim()
      : ''
    if (name) personas.push({ name })
  }
  if (!personas.length) return { ...UNPARSED_PERSONAS }
  const count = typeof rec.count === 'number' && rec.count > 0 ? rec.count : personas.length
  return { count, personas, parsed: true }
}

export function personaInitials(name: string): string {
  const trimmed = name.trim()
  if (!trimmed) return '?'
  const parts = trimmed.split(/[\s._-]+/).filter(Boolean)
  if (parts.length >= 2) {
    return `${parts[0]![0] || ''}${parts[1]![0] || ''}`.toUpperCase()
  }
  return trimmed.slice(0, 2).toUpperCase()
}

File: src/swarm/core/persona_parse.py
"""Static parse of openai-agents persona definitions.

Never executes blueprint source. Used by the blueprints API and team visual
roster (REQ-81 / #433) so the UI can show how many agents a team recipe
declares.
"""

from __future__ import annotations

import ast
from typing import Any

PERSONA_CTORS = frozenset({"Agent", "make_agent", "_make_agent"})

UNPARSED: dict[str, Any] = {"count": 1, "personas": [], "parsed": False}


def parse_openai_agent_personas(source: str | None) -> dict[str, Any]:
    """Return declared personas from Python/API source without executing it.

    Detects ``Agent(...)``, ``make_agent(...)``, and ``_make_agent(...)``
    (software-dev / Chatty-style helpers). A name is taken from a string
    ``name=`` keyword or the first positional string. Variable names are
    skipped (not invented). Unknown / unparsable source yields count 1 and
    an empty persona list.
    """
    if not isinstance(source, str) or not source.strip():
        return dict(UNPARSED)
    try:
        tree = ast.parse(source)
    except (SyntaxError, ValueError, TypeError):
        return dict(UNPARSED)

    names: list[str] = []
    seen: set[str] = set()
    for node in ast.walk(tree):
        if not isinstance(node, ast.Call):
            continue
        if _call_name(node) not in PERSONA_CTORS:
            continue
        name = _literal_name(node)
        if not name or name in seen:
            continue
        seen.add(name)
        names.append(name)

    if not names:
        return dict(UNPARSED)
    return {
        "count": len(names),
        "personas": [{"name": item} for item in names],
        "parsed": True,
    }


def serialize_personas(parsed: dict[str, Any] | None) -> dict[str, Any]:
    """Public JSON shape for API / SPA."""
    data = parsed if isinstance(parsed, dict) else UNPARSED
    personas = [
        {"name": str(row["name"])}
        for row in (data.get("personas") or [])
        if isinstance(row, dict) and str(row.get("name") or "").strip()
    ]
    if not personas:
        return {"count": 1, "personas": [], "parsed": False}
    return {"count": len(personas), "personas": personas, "parsed": True}


def _call_name(node: ast.Call) -> str | None:
    func = node.func
    if isinstance(func, ast.Name):
        return func.id
    if isinstance(func, ast.Attribute):
        return func.attr
    return None


def _literal_name(node: ast.Call) -> str | None:
    for keyword in node.keywords:
        if keyword.arg == "name":
            text = _string_constant(keyword.value)
            if text:
                return text
    if node.args:
        return _string_constant(node.args[0])
    return None


def _string_constant(node: ast.AST) -> str | None:
    if isinstance(node, ast.Constant) and isinstance(node.value, str):
        text = node.value.strip()
        return text or None
    return None
